csharp/src/Microsoft.ML.OnnxTraining/NativeMethodsTraining.cs

```diff
@@ -0,0 +1,306 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Runtime.InteropServices;
+using System.Text;
+
+namespace Microsoft.ML.OnnxTraining
+{
+    /// <summary>
+    /// Enumerations must match the same values as those in orttraining_c_api.h
+    /// </summary>
+    #region Enumreations
+
+    public enum OrtTrainingOptimizer
+    {
+        ORT_TRAINING_OPTIMIZER_SGD = 0
+    };
+
+    public enum OrtTrainingLossFunction
+    {
+        ORT_TRAINING_LOSS_FUNCTION_SOFTMAXCROSSENTROPY = 0
+    };
+
+    public enum OrtTrainingStringParameter
+    {
+        ORT_TRAINING_MODEL_PATH = 0,
+        ORT_TRAINING_LOG_PATH = 1,
+        ORT_TRAINING_INPUT_LABELS = 2,
+        ORT_TRAINING_OUTPUT_PREDICTIONS = 3,
+        ORT_TRAINING_OUTPUT_LOSS = 4
+    };
+
+    public enum OrtTrainingLongParameter
+    {
+        ORT_TRAINING_NUM_TRAIN_STEPS = 0,
+        ORT_TRAINING_TRAIN_BATCH_SIZE = 1,
+        ORT_TRAINING_EVAL_BATCH_SIZE = 2,
+        ORT_TRAINING_EVAL_PERIOD = 3,
+        ORT_TRAINING_DISPLAY_LOSS_STEPS = 4
+    };
+
+    public enum OrtTrainingNumericParameter
+    {
+        ORT_TRAINING_LEARNING_RATE = 0
+    };
+
+    public enum OrtTrainingBooleanParameter
+    {
+        ORT_TRAINING_USE_GIST = 0,
+        ORT_TRAINING_USE_CUDA = 1,
+        ORT_TRAINING_USE_PROFILER = 2,
+        ORT_TRAINING_USE_TENSORBOARD = 3,
+        ORT_TRAINING_IS_PERFTEST = 4,
+        ORT_TRAINING_SHUFFLE_DATA = 5
+    };
+
+    #endregion
+
+    [StructLayout(LayoutKind.Sequential)]
+    public struct OrtTrainingApiBase
+    {
+        public IntPtr GetApi;
+        public IntPtr GetVersionString;
+    };
+
+    [UnmanagedFunctionPointer(CallingConvention.StdCall)]
+    delegate void OrtErrorFunctionCallback(UIntPtr nCount, IntPtr /* (OrtValueCollection*) */ colVal); 
```

Comment:
We will need to review the functions that are exported standalone. We can not implement versioning of them or deprecate them as if when they are a part of the API structure.

Reply:
If you have a better way to implement a call-back from the C++ to C#, I am all ears.

Reply:
Let me know when you are done with your review for I have a number of updates ready.